hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/lock/OmRequestGatekeeper.java

```diff
@@ -0,0 +1,242 @@
+    }
+
+    try {
+      acquireLocks(locks);
+      lockStatsBegin(omLockObject.getLockStats(), Time.monotonicNowNanos(), startTime);
+    } catch (InterruptedException e) {
+      locks.clear();
+      Thread.currentThread().interrupt();
+      throw new OMException("waiting for locks is interrupted", OMException.ResultCodes.INTERNAL_ERROR);
+    } catch (TimeoutException e) {
+      locks.clear();
+      throw new OMException("Unable to get locks, timeout occurred", OMException.ResultCodes.TIMEOUT);
+    }
+    return omLockObject;
+  }
+
+  /*
+  Optional: If we want more diagnostic info on the type of lock that failed to be acquired (volume, bucket, or key),
+  We can make the parameter a list of objects that wrap the Lock with information about its type.
+  Note that logging the specific volume, bucket or keys this lock was trying to acquire is not helpful and
+  misleading because collisions within the stripe lock might mean we are blocked on a request for a completely
+  different part of the namespace.
+  Obtaining the thread ID that we were waiting on would be more useful, but there is no easy way to do that.
+   */
+  private void acquireLocks(List<Lock> locks) throws TimeoutException, InterruptedException {
+    List<Lock> acquiredLocks = new ArrayList<>(locks.size());
+    for (Lock lock: locks) {
+      if (lock.tryLock(LOCK_TIMEOUT_DEFAULT, TimeUnit.MILLISECONDS)) {
+        try {
+          acquiredLocks.add(lock);
+        } catch (Throwable e) {
+          // We acquired this lock but were unable to add it to our acquired locks list.
+          lock.unlock();
+          releaseLocks(acquiredLocks);
+          throw e;
+        }
+      } else {
+        releaseLocks(acquiredLocks);
+        throw new TimeoutException("Failed to acquire lock after the given timeout.");
+      }
+    }
+  }
+
+  public void unlock(OmLockObject lockObject) {
+    releaseLocks(lockObject.getLocks());
+    lockStatsEnd(lockObject.getLockStats(), lockObject.getLockStatType());
+    lockObject.getLocks().clear();
+  }
+
+  private void releaseLocks(List<Lock> locks) {
+    ListIterator<Lock> reverseIterator = locks.listIterator(locks.size());
+    while (reverseIterator.hasPrevious()) {
+      Lock lock = reverseIterator.previous();
+      lock.unlock();
+    }
+  }
+
+  private static void lockStatsBegin(OmLockStats lockStats, long endTime, long startTime) {
+    lockStats.add(endTime - startTime, OmLockStats.Type.WAIT);
+    lockStats.setLockStartTime(endTime);
+  }
+
+  private static void lockStatsEnd(OmLockStats lockStats, OmLockStats.Type type) {
+    if (lockStats.getLockStartTime() > 0) {
+      lockStats.add(Time.monotonicNowNanos() - lockStats.getLockStartTime(), type);
+    }
+  }
+
+  /**
+   * Lock information after taking locks.
+   */
+  public static class OmLockObject {
+    private final OmLockInfo omLockInfo;
+    private final List<Lock> locks = new ArrayList<>();
+    private final OmLockStats lockStats = new OmLockStats();
```

Comment:
Are these fields intended to be static?

Reply:
No, as this is referred for each request, and can not be static.